sections/errorhandling/correlationid.md

```diff
@@ -0,0 +1,35 @@
+# Correlation ID: help your logs tell you a story and give your error a context
+
+### One Paragraph Explainer
+
+Correlation ID is one of the best problem-solving patterns. It lets you link log records, even if they belong to different services. If your system consumes other services and is itself a producer service, adding a correlaction ID is a must. By this pattern, your transaction logs can become into a story that tells itself by filtering your logs with a specific correlation ID, instead of try linking the cross-services transaction logs to each other by yourself. It can save your day when a process including 20 different microservices throws an exception in one of them, and you have no idea where did the problem start across the flow.
+
+<br/>
+
+### Code Example: passing the correlation ID between services on the requet http context
+Here is an example of using [express-http-context](https://www.npmjs.com/package/express-http-context) library as part of a middleware to set the forwarded correlation ID on the http context after it was extracted from the request. 
+If the correlation ID doesn't exist on the request, since this may be the first request in the requests chain, it's being generated.
+
+```javascript
+const httpContext = require('express-http-context');
```

Comment:
Node now has a built-in mechanism, async-local-storage which is also not express specific (we try to move away from being too express-centric). Shall we use this?

Reply:
Very interesting - I didn't know about this ability.
I'm not sure it should completely replace the current example; async-local-storage is useful only for the async functions use case you mentioned.
The multiple-microservices use case can't be solved by that async-local-storage solution...

Reply:
Using 'cls-hooked' and 'cls-rtracer' works very well for the use case you are describing. With multiple microservices one just has to add an http header with the correlationId on outgoing calls so that it's picked by the next MS.

Those plugin uses the `async-local-storage` api under the hood and provide middleware for well-known frameworks like express, fastify, hapi, koa

Reply:
@kevynb You sure that cls-hooked is using async-local-storage? Why use this and not async-local-storage directly?

Reply:
@goldbergyoni 
Check my last commits - cls-rtracer totally simplifies the process.
I changed the whole PR

Reply:
@goldbergyoni Yes, they say it on the repo homepage and you can find it if you dig into the code